src/rdkafka_queue.c

```diff
@@ -361,7 +361,8 @@ rd_kafka_op_t *rd_kafka_q_pop_serve (rd_kafka_q_t *rkq, int timeout_ms,
 
                         /* No op, wait for one */
                         pre = rd_clock();
-			if (cnd_timedwait_ms(&rkq->rkq_cond,
+			if (timeout_ms == 0 ||
```

Comment:
I think we can simply unlock and return NULL if timeout_ms is 0 instead of trying again, that's what will happen below at the `while (timeout_ms != ..NOWAIT)` anyway,

Put the check above the `pre = ..` assignment.

Reply:
@edenhill, actually that's what this already does, because in the case cnd_timedwait() returns timedout we just unlock and return null.  That said we could avoid the call to clock_gettime() which would save us a couple of microseconds while holding a lock.  So I'll make this a separate check.